File: problems/0801-0900/0809_least-weight-subtree-joining-three-nodes/solution_rmq_lca.ts
function leastSubtreeWeight(edges: number[][], queries: number[][]): number[] {
    const n = edges.length + 1;
    const head: Int32Array = new Int32Array(n + 1).fill(-1);
    const nxt = new Int32Array(2 * (n - 1));
    const to = new Int32Array(2 * (n - 1));
    const wt = new Int32Array(2 * (n - 1));
    let cnt = 0;
    for (const [u, v, w] of edges) {
        to[cnt] = v;
        wt[cnt] = w;
        nxt[cnt] = head[u];
        head[u] = cnt++;
        to[cnt] = u;
        wt[cnt] = w;
        nxt[cnt] = head[v];
        head[v] = cnt++;
    }

    // Root at 0 and walk an Euler tour iteratively, so deep chains cannot
    // overflow the call stack. Every node enters the tour at its first visit
    // and re-enters each time a child's subtree closes, giving 2n - 1
    // entries; first[v] is v's earliest slot in that sequence.
    const depth = new Int32Array(n);
    const dist = new Float64Array(n);
    const parent = new Int32Array(n);
    parent[0] = -1;
    const first = new Int32Array(n);
    const it: Int32Array = head.slice();
    const tour = new Int32Array(2 * n - 1);
    let m = 0;
    tour[m++] = 0;
    const stack: number[] = [0];
    while (stack.length > 0) {
        const u = stack[stack.length - 1];
        const e = it[u];
        if (e === -1) {
            stack.pop();
            if (stack.length > 0) tour[m++] = stack[stack.length - 1];
            continue;
        }
        it[u] = nxt[e];
        const v = to[e];
        if (v !== parent[u]) {
            parent[v] = u;
            depth[v] = depth[u] + 1;
            dist[v] = dist[u] + wt[e];
            first[v] = m;
            tour[m++] = v;
            stack.push(v);
        }
    }

    // Sparse table: table[k][i] is the shallowest node over the 2^k tour
    // entries from i - the range argmin under depth comparison.
    let log = 1;
    while (1 << log <= m) log++;
    const table: Int32Array[] = [];
    table.push(tour.slice());
    for (let k = 1; k < log; k++) {
        const prev = table[k - 1];
        const half = 1 << (k - 1);
        const len = m - (1 << k) + 1;
        const cur = new Int32Array(len);
        for (let i = 0; i < len; i++) {
            const a = prev[i];
            const b = prev[i + half];
            cur[i] = depth[a] <= depth[b] ? a : b;
        }
        table.push(cur);
    }

    const lca = (x: number, y: number): number => {
        let l = first[x];
        let r = first[y];
        if (l > r) [l, r] = [r, l];
        const k = 31 - Math.clz32(r - l + 1);
        const a = table[k][l];
        const b = table[k][r - (1 << k) + 1];
        return depth[a] <= depth[b] ? a : b;
    };
    const distance = (x: number, y: number): number => dist[x] + dist[y] - 2 * dist[lca(x, y)];

    // The minimal subtree joining a, b, c is the union of the three paths,
    // each edge lying on exactly two of them.
    return queries.map(([a, b, c]) => (distance(a, b) + distance(b, c) + distance(c, a)) / 2);
}

File: problems/3501-3600/3553_least-weight-subtree-joining-three-nodes/solution_rmq_lca.go
import "math/bits"

func leastSubtreeWeight(edges [][]int, queries [][]int) []int64 {
	n := len(edges) + 1
	type node struct{ v, w int }
	adj := make([][]node, n)
	for _, e := range edges {
		adj[e[0]] = append(adj[e[0]], node{e[1], e[2]})
		adj[e[1]] = append(adj[e[1]], node{e[0], e[2]})
	}

	// Root at 0 and walk an Euler tour iteratively, so deep chains cannot
	// overflow the call stack. Every node enters the tour at its first visit
	// and re-enters each time a child's subtree closes, giving 2n - 1
	// entries; first[v] is v's earliest slot in that sequence.
	depth := make([]int, n)
	dist := make([]int64, n)
	parent := make([]int, n)
	parent[0] = -1
	first := make([]int, n)
	it := make([]int, n)
	tour := make([]int, 0, 2*n-1)
	stack := []int{0}
	tour = append(tour, 0)
	for len(stack) > 0 {
		u := stack[len(stack)-1]
		if it[u] < len(adj[u]) {
			e := adj[u][it[u]]
			it[u]++
			if e.v != parent[u] {
				parent[e.v] = u
				depth[e.v] = depth[u] + 1
				dist[e.v] = dist[u] + int64(e.w)
				first[e.v] = len(tour)
				tour = append(tour, e.v)
				stack = append(stack, e.v)
			}
		} else {
			stack = stack[:len(stack)-1]
			if len(stack) > 0 {
				tour = append(tour, stack[len(stack)-1])
			}
		}
	}
	m := len(tour)

	// Sparse table: table[k][i] is the shallowest node over the 2^k tour
	// entries from i - the range argmin under depth comparison.
	log := 1
	for 1<<log <= m {
		log++
	}
	table := make([][]int, log)
	table[0] = tour
	for k := 1; k < log; k++ {
		prev := table[k-1]
		half := 1 << (k - 1)
		length := m - (1 << k) + 1
		cur := make([]int, length)
		for i := 0; i < length; i++ {
			if depth[prev[i+half]] < depth[prev[i]] {
				cur[i] = prev[i+half]
			} else {
				cur[i] = prev[i]
			}
		}
		table[k] = cur
	}

	lca := func(x, y int) int {
		l, r := first[x], first[y]
		if l > r {
			l, r = r, l
		}
		k := bits.Len(uint(r-l+1)) - 1
		a, b := table[k][l], table[k][r-(1<<k)+1]
		if depth[a] > depth[b] {
			a = b
		}
		return a
	}
	distance := func(x, y int) int64 {
		return dist[x] + dist[y] - 2*dist[lca(x, y)]
	}

	// The minimal subtree joining a, b, c is the union of the three paths,
	// each edge lying on exactly two of them.
	answer := make([]int64, len(queries))
	for j, q := range queries {
		answer[j] = (distance(q[0], q[1]) + distance(q[1], q[2]) + distance(q[2], q[0])) / 2
	}
	return answer
}
